lib/ddtrace/context.rb

```diff
@@ -90,6 +134,66 @@ def sampled?
       end
     end
 
+    # Returns ids of all spans which can be considered as local, partial roots
+    # from a partial flush perspective. Also returns the span IDs which have
+    # been marked as non flushable, and which should be kept.
+    def partial_roots
+      return nil unless @current_span
+
+      marked_ids = Hash[([@current_span.span_id] + @current_span.parent_ids).map { |id| [id, true] }]
+      roots = []
+      @trace.each do |span|
+        # Skip if span is one of the parents of the current span.
+        next if marked_ids.key? span.span_id
+        # Skip if the span is not one of the parents of the current span,
+        # and its parent is not either. It means it just can't be a local, partial root.
+        next unless marked_ids.key? span.parent_id
+
+        roots << span.span_id
+      end
+      [roots, marked_ids]
+    end
+
+    # Return a hash containting all sub traces which are candidates for
+    # a partial flush.
+    def partial_roots_spans
+      roots, marked_ids = partial_roots()
+      return nil unless roots
+
+      roots_spans = Hash[roots.map { |id| [id, []] }]
+      unfinished = {}
+      @trace.each do |span|
+        ids = [span.span_id] + span.parent_ids()
+        ids.reject! { |id| marked_ids.key? id }
+        ids.each do |id|
+          if roots_spans.key?(id)
+            unfinished[id] = true unless span.finished?
```

Comment:
I think we can simplify this (and avoid unnecessary lookups) a little bit by doing:
```rb
root_spans.delete(id) unless span.finished?
```
That way we can get rid of the`unfinished` hash and `roots_spans.reject! { |id| unfinished.key? id }` altogether.

Reply:
Will try this for sure, sounds good.